db/context.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+// txnContextKey is the key type for transaction context values.
+type txnContextKey struct{}
+
+// explicitTxn is a transaction that is managed outside of a db operation.
+type explicitTxn struct {
+	datastore.Txn
+}
+
+func (t *explicitTxn) Commit(ctx context.Context) error {
+	return nil // do nothing
+}
+
+func (t *explicitTxn) Discard(ctx context.Context) {
+	// do nothing
+}
+
+// transactionDB is a db that can create transactions.
+type transactionDB interface {
+	NewTxn(context.Context, bool) (datastore.Txn, error)
+}
+
+// ensureContextTxn ensures that the returned context has a transaction.
+//
+// If a transactions exists on the context it will be made explicit,
+// otherwise a new implicit transaction will be created.
+func ensureContextTxn(ctx context.Context, db transactionDB, readOnly bool) (context.Context, error) {
```

Comment:
suggestion: There is a lot of `ensureContextTxn` followed by `mustGetContextTxn`. To avoid having to type cast everytime, I would suggest returning `datastore.Txn` here. 
```go
func ensureContextTxn(ctx context.Context, db transactionDB, readOnly bool) (context.Context, datastore.Txn, error)
```